R/ctwas_wrapper.R

```diff
@@ -165,123 +147,60 @@ select_ctwas_weights <- function(weight_db_files, conditions = NULL,
-        model_selection[k] <- list(NULL)
-        names(conditions)[match(k, conditions)] <- "non_imputable"
-      }
-    }
-  }
-  # run ctwas_select on all conditions first to keep list structure the same, regardless of imputable status.
-  ctwas_select_result <- ctwas_select(combined_all_data, conditions, names(conditions), max_var_selection, variable_name_obj, susie_obj)
-  weights <- consolidate_weights_list(combined_all_data, conditions, variable_name_obj, twas_weights_table)
-  # assign the weights_selection as NULL in ctwas_select_result #
-  if (length(non_imp_conditions) >= 1) {
-    for (unimp_conds in non_imp_conditions) {
-      ctwas_select_result[[unimp_conds]][["variant_selection"]] <- rep(NA, length(ctwas_select_result[[unimp_conds]][["variant_selection"]]))
-    }
+  
+  
+  ## load twas weights
+  twas_data_combined <- load_twas_weights(weight_db_file, conditions=NULL, variable_name_obj=variable_name_obj,
+                                                        twas_weights_table = twas_weights_table)
+  
+  weights <- twas_data_combined$weights
+  if (is.null(contexts)) {contexts <- names(weights)}
+  
+  #get gene name for the weight file 
+  gene <- unique(twas_data_combined$susie_results[[1]]$region_info$region_name)
+  
+  ## we first select best model, determine imputable contexts, then select variants based on susie obj output
+  model_selection <- pick_best_model(twas_data_combined, contexts, min_rsq_threshold, p_val_cutoff) 
+  # Determine Imputable Contexts  
+  imputable_contexts <- c()
+  for (contx in contexts){
+    if (!is.null(model_selection[[contx]])) imputable_contexts <- c(imputable_contexts, contx)
   }
+  # if (length(imputable_contexts)==0){
+  #   model_selection$region_imputable <- FALSE
+  #   print(paste0("No model meets the p_value threshold and R-squared minimum in all contexts. All contexts for gene ", gene, " at region ", ld_region, " is not imputable. Weight input: ", weight_db_file, ". "))
+  #   # return all results regardless of imputation status. 
+  #   # return(list(susie_results = NULL, model_selection = model_selection, weights = NULL, region_info = twas_data_combined$susie_results[[1]]$region_info))
+  # } else{
+  #   model_selection$region_imputable <- TRUE
+  #   # run top signal selection for all imputable contexts first to keep list structure the same, regardless of imputable status.
+  # }
+  
+  # select variants 
+  names(contexts) <- rep("non_imputable", length(contexts))
+  names(contexts)[which(contexts %in% imputable_contexts)] <- "imputable"
+  ctwas_select_result <- ctwas_select(twas_data_combined, contexts, names(contexts), max_var_selection, variable_name_obj, susie_obj)
+  
   # return results
   return(list(
     susie_results = ctwas_select_result, model_selection = model_selection,
-    weights = weights, region_info = region_info
+    weights = weights, region_info=twas_data_combined$susie_results[[1]]$region_info
```

Comment:
I think `region_info` is in the original weight database. It's a bit confusing to read it as `susie_result` because we don't always have to run `susie` at all to get the original weight DB.

Reply:
yes, indeed, I can replace that as gene=gene, the main purpose is just to identify which gene this rds is for.